lib/Conversion/ExportVerilog/LegalizeNames.cpp

```diff
@@ -41,6 +41,39 @@ StringRef NameCollisionResolver::getLegalName(StringRef originalName) {
   return legalizeName(originalName, usedNames, nextGeneratedNameID);
 }
 
+//===----------------------------------------------------------------------===//
+// FieldNameResolver
+//===----------------------------------------------------------------------===//
+
+void FieldNameResolver::setRenamedFieldName(StringAttr fieldName,
+                                            StringAttr newFieldName) {
+  renamedFieldNames[fieldName] = newFieldName;
+  usedFieldNames.insert(newFieldName);
+}
+
+StringAttr FieldNameResolver::getRenamedFieldName(StringAttr fieldName) {
+  auto it = renamedFieldNames.find(fieldName);
+  if (it != renamedFieldNames.end())
+    return it->second;
+
+  // If a field name is not verilog name or used already, we have to rename it.
+  bool hasToBeRenamed = !sv::isNameValid(fieldName.getValue()) ||
+                        usedFieldNames.count(fieldName.getValue());
+
+  if (!hasToBeRenamed) {
+    setRenamedFieldName(fieldName, fieldName);
```

Comment:
This identity storage seems inefficient.  Is there a way to avoid it?

Reply:
I think this is affordable because this function is called only for struct fields, and the key and element are `StringAttr` so they don't allocate extra memory for strings. Also I'm using `renamedFieldNames` as memoization to avoid checking validity of names multiple times so I think it is simpler to store identity as well.